eva/optimizer/rules/rules.py

```diff
@@ -1262,6 +1262,21 @@ def apply(self, before: LogicalVectorIndexScan, context: OptimizerContext):
         yield after
 
 
+"""
+Rules to optimize Ray.
+"""
+
+
+def get_ray_env_dict():
+    # Get the highest GPU id and expose all GPUs that have id lower than
+    # the max id.
+    if len(Context().gpus) > 0:
+        max_gpu_id = max(Context().gpus) + 1
+        return {"CUDA_VISIBLE_DEVICES": ",".join([str(n) for n in range(max_gpu_id)])}
```

Comment:
Shall we use the GPU ids specified in the eva.yml?

Reply:
I assume you asked why I take max here?

I was thinking of a case that there may exist 3 GPUs [0, 1, 2] and the user wants to use [1, 2]. If we only do `CUDA_VISIBLE_DEVICES=1,2`, the device id on the pytorch side will become [0, 1] inside Ray. Then, it will error out when we do `.to(cuda:2`. 

So, I am thinking to expose all GPUs that are within the range of max id. And this is just to maintain a legit GPU id, but the system access to GPU will still be governed by `Context().gpus` from user setting, so system won't access GPU[0]. 

Let me know if that makes sense.

Reply:
Get it.

Reply:
Just come to my mind. By default, all GPUs are exposed to ray. It seems that is what we want. If so, can we just not set `CUDA_VISIBLE_DEVICES`?